### DISCUSSION TITLE
Run test suite without having to build curl (against an installed binary)

### DISCUSSION THREAD

**Discussion by samueloph** [curl/curl#11073] [Ideas]
I would like to be able to run curl's tests against a currently installed binary of curl without having to build curl itself from source (`make`).

It doesn't seem like it's possible to call `runtests.pl` or `make test-nonflaky` without also having to `make` and `make test`, `make test` is fine (builds test artifacts) but `make` will build the whole curl in order to get libcurl (for libauthretry). 

My understanding is that `make` is needed as `libcurl.la` is a required target for `libauthretry` (from `make test`), I wonder if it would be possible to tweak the makefile to allow for libauthretry to link to the system's libcurl.so (different behavior based on an environment variable, maybe) and thus avoid the need to build everything when testing.

```
make[3]: *** No rule to make target '../../lib/libcurl.la', needed by 'libauthretry'.  Stop.
```

I do understand this request has a very low priority due to workaround being possible (one can build curl in order to run the tests) and the impact being small, but I wanted to at least make a record of it.

I am hopping that the ability to more easily run curl's tests against an installed binary without having to build curl will be helpful for others as well, eg.: users could pull a more recent test suite from curl and run against their old versions to look for possible issues (that is, without having to build curl). CI tests can install pre-compiled binaries of curl and run all of the unit tests without having to rebuild it just for the sake of linking libauthretry (after all the test is done against the pre-compiled binary), this is the case of what we started doing in Debian, where the CI runs whenever any dependency changes.

I hope the request makes sense.

**Comment by bagder** [comment]
It mostly already supports this, using the `-c` option to runtests:
~~~
$ cd tests
$ make
$ ./runtests.pl -n -c /usr/bin/curl
...
~~~

I noticed just now that I needed `-n` to switch off valgrind as otherwise it fails with an error, but that seems like a temporary bug.

**Reply by bagder** [reply]
I filed #11074 for that.

**Reply by samueloph** [reply]
> It mostly already supports this, using the `-c` option to runtests:
> 
> ```
> $ cd tests
> $ make
> $ ./runtests.pl -n -c /usr/bin/curl
> ...
> ```

The problem is the requirement to run "make", which will build the whole curl even if I intend to run the tests with `-c` and `-vc` afterwards.

The request is to be able to run the tests without having to build curl, building the test artifacts is fine, but for the tests, libauthretry needs to link to libcurl (the one built, not the system one), which ends up creating a requirement to build curl.

**Reply by bagder** [reply]
> The problem is the requirement to run "make", which will build the whole curl even if I intend to run the tests with -c and -vc afterwards.

That's not "the problem". The test suite itself builds what it needs to run, and it needs parts of the curl code built.

> for the tests, libauthretry needs to link to libcurl (the one built, not the system one), which ends up creating a requirement to build curl.

Sure, but I don't consider that a problem. It might build a file or two too many in that case.

I think I will put that under the "mostly" statement I made. The test suite is currently not designed to use and run with a custom libcurl the same way it can run with a custom curl binary. Room for improvement. I've personally not had the need.
